forge/db/models/Team.js

```diff
@@ -354,6 +354,87 @@ module.exports = {
+                            code: 'device_limit_reached',
+                            error: 'Device limit reached',
+                            limit: targetDeviceLimit,
+                            count: currentDeviceCount
+                        })
+                    }
+
+                    const currentInstanceCountsByType = await this.instanceCountByType()
+                    // console.log(currentInstanceCountsByType)
+                    const targetInstanceLimits = {}
+                    for (const instanceType of Object.keys(currentInstanceCountsByType)) {
+                        // console.log('checking', instanceType, teamType.properties.instances[instanceType])
+                        if (!teamType.getInstanceTypeProperty(instanceType, 'active', false)) {
+                            targetInstanceLimits[instanceType] = 0
+                        } else {
+                            targetInstanceLimits[instanceType] = teamType.getInstanceTypeProperty(instanceType, 'limit', -1)
+                        }
+                        if (targetInstanceLimits[instanceType] !== -1 && targetInstanceLimits[instanceType] < currentInstanceCountsByType[instanceType]) {
+                            errors.push({
+                                code: 'instance_limit_reached',
+                                error: `Instance type ${instanceType} limit reached`,
+                                type: instanceType,
+                                limit: targetInstanceLimits[instanceType],
+                                count: currentInstanceCountsByType[instanceType]
+                            })
+                        }
+                    }
+                    if (errors.length > 0) {
+                        const message = errors.map(err => `${err.error} (current: ${err.count}, limit: ${err.limit})`).join(', ')
+                        const err = new Error(`Unable to change team type: ${message}`)
+                        err.code = 'invalid_request'
+                        err.errors = errors
+                        throw err
+                    }
+                    // console.log(`members: ${currentMemberCount} / ${targetMemberLimit}`)
+                    // console.log(`devices: ${currentDeviceCount} / ${targetDeviceLimit}`)
+                    // for (const instanceType of Object.keys(currentInstanceCountsByType)) {
+                    //     console.log(`instance[${instanceType}] ${currentInstanceCountsByType[instanceType]} / ${targetInstanceLimits[instanceType]}`)
+                    // }
+                },
+                /**
+                 * Update the team type.
+                 *
+                 * When running with EE, this function is replaced via ee/lib/billing/Team.js
+                 * to add additional checks related to billing
+                 */
+                updateTeamType: async function (teamType) {
+                    await this.setTeamType(teamType)
```

Comment:
I'm slightly surprised that this method (updateTeamType) doesn't call `checkTeamTypeUpdateAllowed` and raise an exception if the change can't be made. Feels like it opens up the potential of team type being updated to an invalid state if `checkTeamTypeUpdateAllowed` is forgotten to be called first.